tests/queue/tests.c

```diff
@@ -0,0 +1,411 @@
+/* Copyright 2014 Nodalink EURL
+
+ * This file is part of Butterfly.
+ *
+ * Butterfly is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as published
+ * by the Free Software Foundation.
+ *
+ * Butterfly is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <glib.h>
+#include <string.h>
+#include <packetgraph/packetgraph.h>
+#include "brick-int.h"
+#include "packets.h"
+#include "utils/mempool.h"
+#include "utils/bitmask.h"
+#include "utils/mac.h"
+#include "collect.h"
+
+#define CHECK_ERROR(error) do {                 \
+               if (error) {                     \
+                       pg_error_print(error);   \
+		       pg_error_free(error);    \
+		}                               \
+               g_assert(!error);                \
+       } while (0)
+
+static void test_queue_lifecycle(void)
+{
+	struct pg_error *error = NULL;
+	struct pg_brick *brick;
+
+	brick = pg_queue_new("test_queue", 0, &error);
+	CHECK_ERROR(error);
+	brick = pg_queue_new("test_queue", 1, &error);
+	CHECK_ERROR(error);
+
+	pg_brick_decref(brick, &error);
+	CHECK_ERROR(error);
+}
+
+static void test_queue_friend(void)
+{
+	struct pg_error *error = NULL;
+	struct pg_brick *q1, *q2, *q3;
+
+	q1 = pg_queue_new("q1", 1, &error);
+	CHECK_ERROR(error);
+	q2 = pg_queue_new("q2", 1, &error);
+	CHECK_ERROR(error);
+	q3 = pg_queue_new("q3", 1, &error);
+	CHECK_ERROR(error);
+
+	/* classic scenario */
+	g_assert(!pg_queue_get_friend(q1));
+	g_assert(!pg_queue_get_friend(q2));
+	g_assert(!pg_queue_friend(q1, q2, &error));
+	CHECK_ERROR(error);
+	g_assert(pg_queue_get_friend(q1) == q2);
```

Comment:
the first assert is not useful, because pg_queue_are_friend do exactly the same thing.